### PR TITLE
Reorganization of the repo

### PR DESCRIPTION
**Pull request by W95Psp**
This PR reorganizes and renames the different components of this project:
- `thir-export` is now `circus-frontend`, in the folder `rust-frontend`;
-  `thir-elab` is now`circus-engine`, in the folder `engine`;
-  the directory `cli` contains the Rust crate for the binary of the circus subcommand;
-  the backends live in directories `engine/backends/BACKEND` (with `BACKEND` being `fstar` or `coq` currently)

### CODE REVIEW

**Review by franziskuskiefer — APPROVED** [review]
Thanks, this looks much nicer.
There are things (like the one I commented on) that we want to do as well. But not here.
The structure looks good to me. We'll have to talk about the exact names (also to make them publishable on crates.io). But it's a follow-up as welll.

File: cli/Cargo.toml [hunk 1]
```diff
@@ -6,20 +6,20 @@ edition = "2021"
...
-path = "src/cargo_thir_export.rs"
-name = "cargo-thir-export"
+path = "src/cargo_circus_frontend.rs"
+name = "cargo-circus-frontend"
 
 [[bin]]
 path = "src/cargo_circus.rs"
 name = "cargo-circus"
 
 [[bin]]
```
[the remark this answers is not in the document]
Comment (W95Psp) [reply cli/Cargo.toml:16]: I absolutely agree, I also think it's overkill currently, I'll fix that

**Comment by karthikbhargavan** [comment]
Any reason why `backends` live inside `engine`? Would be nice if we could pull them out so that anyone can build a backend based on engine, no?

**Comment by W95Psp** [comment]
I wanted initially to have a `backend` folder at root, but I was struggling at making opam happy with it, basically... :(

**Comment by W95Psp** [comment]
I've added a `diagnostic` crate for hosting errors, cleaned up the JSON schema export binary in Rust, cleaned up the CLI, removed one binary in favor of a subcommand in `circus`

I think we can merge this!

**Pull request merged**